File: packages/deon-go/options.go
package deon

// DefaultSourceName is what a document is called when the caller does not name it.
const DefaultSourceName = "<memory>"

// ParseOptions are the capabilities and the surroundings of a parse.
//
// Everything is denied by default (specification 9): calling a parser on a piece of text grants
// neither the filesystem nor the network, because a document handed to a library came from somewhere
// unknown. Each capability is an explicit decision.
type ParseOptions struct {
	// SourceName is what the document is called in a diagnostic, and what a relative target resolves
	// against.
	SourceName string

	// Filebase is the directory a relative target resolves against, when it is not the one holding
	// the source.
	Filebase string

	// Resources handed to the evaluator directly, keyed by target. Consulted before any loader, so a
	// document that imports can be evaluated while reaching nothing at all — how the conformance suite
	// runs, and how an editor reads a document it has open but has not saved.
	Resources map[string]string

	// AbsolutePaths maps a logical absolute target onto the host path that holds it (specification 9).
	// Exact keys win; among /prefix/* wildcards the longest wins. It is a property of the target
	// rather than of whoever resolves it.
	AbsolutePaths map[string]string

	// Environment is what a #$NAME reads. It defaults to empty and is never filled in from the process
	// environment: a library that read the ambient environment would make a document mean one thing on
	// one machine and another on the next. The CLI reads the environment and passes it in.
	Environment map[string]string

	// Authorization is a bearer token per exact lowercase hostname — no port, no path, no wildcard.
	Authorization map[string]string

	// Token is the credential parse-link fetches with, and half of a cache identifier. It is not what
	// an importer sends; that comes from a declaration's `with` or from Authorization.
	Token string

	AllowFilesystem bool
	AllowNetwork    bool

	// Cache turns on the on-disk response cache for fetched resources (specification 9). It is off by
	// default: a cache is a performance decision, and a library that wrote to a shared directory
	// without being asked would be making it for the caller.
	Cache bool

	// CacheDuration is how long a cached response stays fresh, in milliseconds; CacheDirectory is where
	// entries are written. Zero values mean the defaults — one hour, and ~/.deon-cache.
	CacheDuration  int
	CacheDirectory string

	// DatasignFiles are the .datasign contracts that type the parsed data, and DatasignMap the root
	// keys each type applies to (specification 14.1). Both are needed: an empty map is the identity,
	// and reading a contract without one to apply would be reading a file for nothing.
	DatasignFiles []string
	DatasignMap   map[string]string

	// Expansion is the budget on code points produced by substitution — interpolation and string
	// spread — before evaluation is refused with DEON_LIMIT_EXCEEDED (specification 11). It guards
	// against a tiny document that doubles a value at each step until it assembles gigabytes. Zero
	// selects DefaultExpansion, because expansion is always bounded: setting it to zero asks for the
	// default rather than for an unbounded evaluation.
	Expansion uint64
}

// DefaultExpansion is the expansion budget a host gets when it names none: 2^26 code points, far past
// any expansion a document has cause to perform and far below what exhausts a host (specification 11).
const DefaultExpansion uint64 = 1 << 26

func (o *ParseOptions) sourceName() string {
	if o.SourceName == "" {
		return DefaultSourceName
	}
	return o.SourceName
}

// StringifyOptions decide how a value is written (specification 12). The zero value is the readable
// default: four spaces, inline values, no generated header or comments.
type StringifyOptions struct {
	Canonical          bool
	Readable           bool
	Indentation        int
	Leaflinks          bool
	LeaflinkLevel      int
	LeaflinkShortening bool
	GeneratedHeader    bool
	GeneratedComments  bool

	set bool
}

// DefaultStringifyOptions returns the readable defaults. Because the zero StringifyOptions cannot
// tell "the caller wants Readable false" from "the caller passed nothing", callers build from here.
func DefaultStringifyOptions() StringifyOptions {
	return StringifyOptions{
		Readable:           true,
		Indentation:        4,
		LeaflinkLevel:      1,
		LeaflinkShortening: true,
		set:                true,
	}
}

func (o StringifyOptions) resolved() StringifyOptions {
	if !o.set && !o.Canonical && !o.Readable && o.Indentation == 0 {
		return DefaultStringifyOptions()
	}
	if o.Indentation == 0 {
		o.Indentation = 4
	}
	return o
}

File: packages/deon-python/source/deon/options.py
"""What a document is allowed to reach, and what it is given.

Nothing is granted by default (specification 9). Calling a parser on a piece of text grants it
neither the filesystem nor the network; each is a decision somebody has to make out loud. This is
what lets a host evaluate a document it did not write.
"""

from __future__ import annotations

from dataclasses import dataclass, field


DEFAULT_SOURCE_NAME = "<memory>"

#: The default expansion budget (specification 11): the number of Unicode code points that
#: substitution may produce before evaluation is stopped. 2^26 is far above anything a document a
#: person wrote assembles, and far below what a billion-laughs blow-up would — a tiny document doubling
#: an interpolation twenty times reaches gigabytes, and bounding the count is what stops it.
DEFAULT_EXPANSION = 2 ** 26


@dataclass
class ParseOptions:
    #: The name the document is known by, and what a diagnostic points at.
    source_name: str = DEFAULT_SOURCE_NAME

    #: What a relative target inside this document resolves against.
    filebase: str = ""

    #: Resources handed to the evaluator directly, keyed by target. Consulted before any loader, so a
    #: document that imports can be evaluated while reaching nothing at all — which is how the
    #: conformance suite runs, and how an editor reads a document it has open but has not saved.
    resources: dict[str, str] = field(default_factory=dict)

    #: Logical absolute targets, mapped to host paths. Exact keys win; among `/prefix/*` wildcards
    #: the longest wins. The mapping is a property of the *target* rather than of whoever resolves
    #: it, so a resource handed over directly maps exactly as one read from a disk would.
    absolute_paths: dict[str, str] = field(default_factory=dict)

    #: The evaluation environment read by `#$NAME`.
    #:
    #: It defaults to empty and it is never filled in from `os.environ`. A library that read the
    #: ambient environment would make a document mean one thing on one machine and another thing on
    #: the next, and would make the conformance suite pass or fail depending on what the developer
    #: running it happened to have exported. The CLI reads `os.environ` and passes it in; the library
    #: does not go looking.
    environment: dict[str, str] = field(default_factory=dict)

    allow_filesystem: bool = False
    allow_network: bool = False

    #: The most Unicode code points substitution — an interpolation `#{…}`, or a string spread — may
    #: produce before evaluation is stopped with `DEON_LIMIT_EXCEEDED` (specification 11). A tiny
    #: document can otherwise assemble gigabytes by interpolation doubling, which is a denial of
    #: service the host has no other way to refuse. Absent or 0 means `DEFAULT_EXPANSION`; a host that
    #: wants a tighter bound names a smaller number.
    expansion: int = 0

    #: A bearer token per exact lowercase hostname. Exact: no port, no path, no wildcard. A
    #: credential is not something to hand out on a prefix match.
    authorization: dict[str, str] = field(default_factory=dict)

    #: The credential `parse_link` fetches with. It is not what an *importer* sends — that comes from
    #: the declaration's `with`, or from `authorization` — and it is the importer's resolved credential,
    #: not this one, that a cache entry is separated by (specification 9).
    token: str = ""

    cache: bool = False
    cache_duration: int = 3_600_000
    cache_directory: str = ""

    #: The `.datasign` contracts that type the parsed data, and the root keys each type applies to
    #: (specification 14.1). Both are needed: an empty map is the identity, and reading a contract
    #: without one to apply would be reading a file for nothing.
    datasign_files: list[str] = field(default_factory=list)
    datasign_map: dict[str, str] = field(default_factory=dict)



@dataclass
class StringifyOptions:
    #: The canonical form is the one two implementations must agree on, character for character
    #: (specification 13). It is not a style; it is an identity.
    canonical: bool = False

    readable: bool = True
    indentation: int = 4

    leaflinks: bool = False
    leaflink_level: int = 1
    leaflink_shortening: bool = True

    generated_header: bool = False
    generated_comments: bool = False
